trin-core/src/portalnet/overlay_service.rs

```diff
@@ -176,96 +241,228 @@ impl OverlayService {
+                // associated database, then look for a radius value there.
+                let node = Node::new(enr, U256::from(u64::MAX));
+                let status = NodeStatus {
+                    state: ConnectionState::Disconnected,
+                    direction: ConnectionDirection::Outgoing,
+                };
+
+                // Attempt to insert the node into the routing table.
+                match service.kbuckets.write().insert_or_update(
+                    &kbucket::Key::from(node_id.clone()),
+                    node,
+                    status,
+                ) {
+                    InsertResult::Failed(reason) => {
+                        warn!(
+                            "[{:?}] Failed to insert bootnode into overlay routing table. Node: {}, Reason {:?}",
+                            service.protocol, node_id, reason
+                        );
+                    }
+                    _ => {
+                        debug!(
+                            "[{:?}] Inserted bootnode into overlay routing table, adding to ping queue. Node {}",
+                            service.protocol, node_id
+                        );
+
+                        // Queue the node in the ping queue.
+                        service.peers_to_ping.insert(node_id);
+                    }
+                }
+            }
+
+            info!("[{:?}] Starting overlay service", overlay_protocol);
             service.start().await;
         });
 
         Ok(request_tx)
     }
 
+    /// The main loop for the overlay service. The loop selects over different possible tasks to
+    /// perform.
+    ///
+    /// Process request: Process an incoming or outgoing request through the overlay.
+    ///
+    /// Process response: Process a response to an outgoing request from the local node. Try to
+    /// match this response to an active request, and send the response or error over the
+    /// associated response channel. Update node state based on result of response.
+    ///
+    /// Ping queue: Ping a node in the routing table to perform a liveness check and to refresh
+    /// information relevant to the overlay network.
+    ///
+    /// Bucket maintenance: Maintain the routing table (more info documented above function).
     async fn start(&mut self) {
-        while let Some(request) = self.request_rx.recv().await {
-            self.process_request(request).await;
+        loop {
+            tokio::select! {
+                Some(request) = self.request_rx.recv() => self.process_request(request),
+                Some(response) = self.response_rx.recv() => {
+                    // Look up active request that corresponds to the response.
+                    let optional_active_request = self.active_outgoing_requests.write().remove(&response.request_id);
+                    if let Some(active_request) = optional_active_request {
+                        // Send response to responder if present.
```

Comment:
I agree that the wording could be more clear. Here the "responder" is the channel which would carry the response from where it was received to the initiator of the request. If the request was initiated internally, i.e. as part of routing table maintenance, the "responder" is `None`, because we only want to process the contents of the response.

I suppose that we could adopt the `discv5` naming and call this the "callback".